src/ucp/core/ucp_mm.c

```diff
@@ -712,52 +714,53 @@ ucp_rndv_frag_malloc_mpools(ucs_mpool_t *mp, size_t *size_p, void **chunk_p)
         goto out;
     }
 
-    *chunk_p = chunk_hdr + 1;
-    ucs_mpool_chunk_get_info(mp->data, *size_p, (ucs_mpool_chunk_t*)*chunk_p,
-		             &elems, &num_elems);
-    ucs_debug("elems %p, num_elems %u", elems, num_elems);
+    num_elems =
+        ucs_mpool_num_elems_per_chunk(mp, (ucs_mpool_chunk_t*)(chunk_hdr + 1),
+                                      *size_p);
 
-    /* payload */
-    /* Need to get default flags from ucp_mem_map_params2uct_flags() */
+    /* payload; need to get default flags from ucp_mem_map_params2uct_flags() */
     mem_params.field_mask = 0;
-    status = ucp_mem_map_common(mpriv->worker->context, NULL,
-                                mpriv->frag_size * num_elems,
-                                mpriv->mem_type,
+    frag_size             = context->config.ext.rndv_frag_size[mem_type];
+    status = ucp_mem_map_common(context, NULL, frag_size * num_elems, mem_type,
                                 ucp_mem_map_params2uct_flags(&mem_params),
                                 1, ucs_mpool_name(mp), &memh);
     if (status != UCS_OK) {
         goto out;
     }
 
-    chunk_hdr->memh             = memh;
-    chunk_hdr->next_frag_offset = memh->address;
+    chunk_hdr->memh          = memh;
+    chunk_hdr->next_frag_ptr = memh->address;
+    *chunk_p                 = chunk_hdr + 1;
 
 out:
     return status;
 }
 
-static inline void
+static void
 ucp_rndv_frag_free_mpools(ucs_mpool_t *mp, void *chunk)
 {
     ucp_rndv_mpool_priv_t *mpriv = ucs_mpool_priv(mp);
-    ucp_mem_desc_t *chunk_hdr;
+    ucp_rndv_frag_mp_chunk_hdr_t *chunk_hdr;
 
-    chunk_hdr = (ucp_mem_desc_t*)chunk - 1;
+    chunk_hdr = (ucp_rndv_frag_mp_chunk_hdr_t*)chunk - 1;
     ucp_mem_unmap_common(mpriv->worker->context, chunk_hdr->memh);
     ucs_free(chunk_hdr);
 }
 
 void ucp_frag_mpool_obj_init(ucs_mpool_t *mp, void *obj, void *chunk)
 {
-    ucp_mem_desc_t *elem_hdr     = obj;
-    ucp_mem_desc_t *chunk_hdr    = ((ucp_mem_desc_t*)chunk - 1);
-    ucp_rndv_mpool_priv_t *mpriv = ucs_mpool_priv(mp);
-    void *next_frag_offset       = chunk_hdr->next_frag_offset;
-
-    elem_hdr->memh              = chunk_hdr->memh;
-    elem_hdr->ptr               = next_frag_offset;
-    chunk_hdr->next_frag_offset = UCS_PTR_BYTE_OFFSET(next_frag_offset,
-                                                      mpriv->frag_size);
+    ucp_rndv_frag_mp_chunk_hdr_t *chunk_hdr = (ucp_rndv_frag_mp_chunk_hdr_t*)chunk - 1;
+    void *next_frag_ptr                     = chunk_hdr->next_frag_ptr;
+    ucp_rndv_mpool_priv_t *mpriv            = ucs_mpool_priv(mp);
+    ucs_memory_type_t mem_type              = mpriv->mem_type;
+    ucp_context_h context                   = mpriv->worker->context;
+    ucp_context_config_t ext                = context->config.ext;
```

Comment:
remove "ext" helper var
if line gets too long, can move it to before line 761 (separate assign from declaration)